arm9/src/list_menu.c

```diff
@@ -0,0 +1,572 @@
+    }
+    if (newCursorPos != NULL)
+    {
+        *newCursorPos = list->cursorPos;
+    }
+    if (newItemsAbove != NULL)
+    {
+        *newItemsAbove = list->itemsAbove;
+    }
+    return -1;
+}
+
+THUMB_FUNC s32 ListMenuTestInput(struct ListMenu * list, const struct ListMenuTemplate * template, u16 cursorPos, u16 itemsAbove, u16 input, u16 *newCursorPos, u16 *newItemsAbove)
+{
+    return ListMenuTestInputInternal(list, template, cursorPos, itemsAbove, FALSE, input, newCursorPos, newItemsAbove);
+}
+
+THUMB_FUNC void ListMenuOverrideSetColors(struct ListMenu * list, u8 cursorPal, u8 fillValue, u8 cursorShadowPal)
+{
+    list->cursorPal = cursorPal;
+    list->fillValue = fillValue;
+    list->cursorShadowPal = cursorShadowPal;
+    list->overrideEnabled = TRUE;
+}
+
+THUMB_FUNC void ListMenuGetCurrentItemArrayId(struct ListMenu * list, u16 * index_p)
+{
+    *index_p = list->cursorPos + list->itemsAbove;
+}
+
+THUMB_FUNC void ListMenuGetScrollAndRow(struct ListMenu * list, u16 * cursorPos_p, u16 * itemsAbove_p)
+{
+    if (cursorPos_p != NULL)
+        *cursorPos_p = list->cursorPos;
+    if (itemsAbove_p != NULL)
+        *itemsAbove_p = list->itemsAbove;
+}
+
+THUMB_FUNC u8 ListMenuGetUnk33(struct ListMenu * list)
+{
+    return list->unk_33;
+}
+
+THUMB_FUNC s32 ListMenuGetValueByArrayId(struct ListMenu * list, s32 index)
+{
+    return list->template.items[index].index;
+}
+
+THUMB_FUNC s32 ListMenuGetTemplateField(struct ListMenu * list, u32 attr)
+{
+    switch (attr)
+    {
+    case 0:
+        return (s32)list->template.moveCursorFunc;
+    case 1:
+        return (s32)list->template.itemPrintFunc;
+    case 2:
+        return (s32)list->template.totalItems;
+    case 3:
+        return (s32)list->template.maxShowed;
+    case 4:
+        break;
+    case 5:
+        return (s32)list->template.header_X;
+    case 6:
+        return (s32)list->template.item_X;
+    case 7:
+        return (s32)list->template.cursor_X;
+    case 8:
+        return (s32)list->template.upText_Y;
+    case 9:
+        return GetFontAttribute(list->template.fontId, 1) + list->template.itemVerticalPadding;
+    case 10:
+        return (s32)list->template.cursorPal;
+    case 11:
+        return (s32)list->template.fillValue;
+    case 12:
+        return (s32)list->template.cursorShadowPal;
+    case 13:
+        return (s32)list->template.lettersSpacing;
+    case 14:
+        return (s32)list->template.itemVerticalPadding;
+    case 15:
+        return (s32)list->template.scrollMultiple;
+    case 16:
+        return (s32)list->template.fontId;
+    case 17:
+        return (s32)list->template.cursorKind;
+    case 18:
+        return (s32)list->template.window;
+    case 19:
+        return (s32)list->template.unk_1C;
+    }
+
+    return -1;
+}
+
+THUMB_FUNC void ListMenuSetTemplateField(struct ListMenu * list, u32 attr, s32 value)
+{
+    switch (attr)
+    {
+    case 0:
+        list->template.moveCursorFunc = (LM_MOVECURSORFUNC)value;
```

Comment:
ngl these casts feel really awkward

Reply:
Can't assign an integer to a pointer type without an explicit cast.

Reply:
I know, it's just the type names feel awkward being in all caps